src/plonk/permutation/verifier.rs

```diff
@@ -0,0 +1,157 @@
+        for c in &self.permutation_product_commitments {
+            transcript
+                .absorb_point(c)
+                .map_err(|_| Error::TranscriptError)?;
+        }
+        Ok(())
+    }
+
+    pub(crate) fn expressions<'a>(
+        &'a self,
+        vk: &'a VerifyingKey<C>,
+        advice_evals: &'a [C::Scalar],
+        l_0: C::Scalar,
+        beta: ChallengeBeta<C::Scalar>,
+        gamma: ChallengeGamma<C::Scalar>,
+        x: ChallengeX<C::Scalar>,
+    ) -> impl Iterator<Item = C::Scalar> + 'a {
+        iter::empty()
+            // l_0(X) * (1 - z(X)) = 0
+            .chain(
+                self.permutation_product_evals
+                    .iter()
+                    .map(move |product_eval| l_0 * &(C::Scalar::one() - product_eval)),
+            )
+            // z(X) \prod (p(X) + \beta s_i(X) + \gamma)
+            // - z(omega^{-1} X) \prod (p(X) + \delta^i \beta X + \gamma)
+            .chain(
+                vk.cs
+                    .permutations
+                    .iter()
+                    .zip(self.permutation_evals.iter())
+                    .zip(self.permutation_product_evals.iter())
+                    .zip(self.permutation_product_inv_evals.iter())
+                    .map(
+                        move |(((p, permutation_evals), product_eval), product_inv_eval)| {
+                            let mut left = *product_eval;
+                            for (advice_eval, permutation_eval) in p
+                                .columns
+                                .iter()
+                                .map(|&column| {
+                                    advice_evals[vk.cs.get_advice_query_index(column, 0)]
+                                })
+                                .zip(permutation_evals.iter())
+                            {
+                                left *= &(advice_eval + &(*beta * permutation_eval) + &gamma);
+                            }
+
+                            let mut right = *product_inv_eval;
+                            let mut current_delta = *beta * &x;
+                            for advice_eval in p.columns.iter().map(|&column| {
+                                advice_evals[vk.cs.get_advice_query_index(column, 0)]
+                            }) {
+                                right *= &(advice_eval + &current_delta + &gamma);
+                                current_delta *= &C::Scalar::DELTA;
+                            }
+
+                            left - &right
+                        },
+                    ),
+            )
+    }
+
+    pub(crate) fn evals(&self) -> impl Iterator<Item = &C::Scalar> {
+        self.permutation_product_evals
+            .iter()
+            .chain(self.permutation_product_inv_evals.iter())
+            .chain(self.permutation_evals.iter().flat_map(|evals| evals.iter()))
+    }
+
+    pub(crate) fn queries<'a>(
+        &'a self,
+        vk: &'a VerifyingKey<C>,
+        x: C::Scalar,
```

Comment:
should this also be `x: ChallengeX<C::Scalar>`?

Reply:
Good catch! I found a different place this was forgotten yesterday, but didn't spot this one.